Plugins/ExaTrkX/include/Acts/Plugins/ExaTrkX/Tensor.hpp

```diff
@@ -0,0 +1,143 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <cmath>
+#include <cstring>
+#include <functional>
+#include <optional>
+#include <ostream>
+
+/// Forward declare cuda stream, to be able to use the header without cuda
+struct CUstream_st;
+using cudaStream_t = CUstream_st *;
+
+namespace Acts {
+
+/// A simple device description struct
+struct Device {
+  enum class Type { eCPU, eCUDA };
+  Type type = Type::eCPU;
+  std::size_t index = 0;
+
+  static Device Cpu() { return {Type::eCPU, 0}; }
+  static Device Cuda(std::size_t index = 0) { return {Type::eCUDA, index}; }
+
+  bool isCpu() const { return type == Type::eCPU; }
+  bool isCuda() const { return type == Type::eCUDA; }
+
+  bool operator==(const Device &) const = default;
+};
+
+/// Easy printout of device
+inline std::ostream &operator<<(std::ostream &os, Device device) {
+  if (device.type == Device::Type::eCPU) {
+    os << "CPU";
+  } else {
+    os << "CUDA(" << device.index << ")";
+  }
+  return os;
+}
+
+/// Capture the context of the execution
+struct ExecutionContext {
+  Acts::Device device{Acts::Device::Type::eCPU};
+  std::optional<cudaStream_t> stream;
+};
+
+namespace detail {
+
+/// This class implements the memory management for the Acts::Tensor
+class TensorMemoryImpl {
+ public:
+  using Deleter = std::function<void(void *)>;
+
+  TensorMemoryImpl(std::size_t nbytes, const ExecutionContext &execContext);
+
```

Comment:
As far as I can tell, you only use `execContext` in the constructor to extract the device type and the stream. Can't this entire `TensorMemoryImpl` class just be `unique_ptr` with a custom deleter, where you determine the deleter in `Tensor<T>` which captures the stream if needed.

Reply:
Yeah, makes sense to use `unique_ptr` indeed.